tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -288,27 +302,102 @@ class UnsortedSegmentSumOp : public OpKernel {
     auto output_flat = output->flat_outer_dims<T>();
 
     auto data_ptr = data.template flat<T>().data();
-    functor::UnsortedSegmentSumFunctor<Device, T, Index>()(
+    functor::UnsortedSegmentBaseFunctor<Device, T, Index,
+                                        unsortedSegmentInitializationOp,
+                                        unsortedSegmentReductionOp>()(
         context, context->template eigen_device<Device>(), output_rows,
         segment_ids.shape(), segment_flat, data.NumElements(), data_ptr,
         output_flat);
   }
 };
 
-#define REGISTER_CPU_UNSORTED_KERNELS(type, index_type)                \
+template <class T>
+inline void unsortedSegmentInitializationOpSum(
```

Comment:
Can you change all these function names to begin with capital letters? i.e. UnsortedSegmentInitializationOpSum

Also, the function name should be some kind of action like "InitializeUnsortedSegmentOpSum"   or if a functor, something like "UnsortedSegmentSumInitOp::Run"